### PR TITLE
resolve atomicity issues on write by not updating metadata collection

### LINKED ISSUE #180: Metadata sequence number out of sync with journal entry sequence number
We have noticed that for our entities using mongodb persistence it was possible for the sequence number in the metadata to be smaller then the journal sequence number.

We do not know why this occurred but we suspect it is because we had 2 entities running with the same persistence id because of a network partition.

When the persistent actor was restarted and and the actor try to write an event we would get a rejection because a journal entry exists but at the same time the sequence number in the metadata would be incremented.

It seems possible from the write message method to have a rejection in the journal and to set the highest sequence number to a value lower then the current highest number.

We could change the SetHighSeqenceId method to an upsert with the Max function to avoid decrementing the max sequence nr in metadata

``` C#
        private async Task SetHighSequenceId(IList<AtomicWrite> messages)
        {
            var persistenceId = messages.Select(c => c.PersistenceId).First();
            var highSequenceId = messages.Max(c => c.HighestSequenceNr);
            var builder = Builders<MetadataEntry>.Filter;
            var filter = builder.Eq(x => x.PersistenceId, persistenceId);

            await _metadataCollection.Value.UpdateOneAsync(filter, Builders<MetadataEntry>.Update.Combine(
                Builders<MetadataEntry>.Update.SetOnInsert(d => d.Id, persistenceId),
                Builders<MetadataEntry>.Update.SetOnInsert(d => d.PersistenceId, persistenceId),
                Builders<MetadataEntry>.Update.Max(d => d.SequenceNr, highSequenceId)), new UpdateOptions {IsUpsert = true});
            
        }
```

Akka version : 1.4.12
Persistence plugin version: 1.4.12

### PR DESCRIPTION
**Pull request by Aaronontheweb**
only update metadata collection upon delete, just like the SQL journals all do - this keeps all writes atomic at the time they're made. Recover data using the highest of two: max seq no from journal or max seq no from metadata.

close #180

### CODE REVIEW

**Review by simonlaroche — COMMENTED** [review]
Looks good.

The only issue I can see is the InsertMany could be within a transaction if there are multiple events to persist.

**Comment by Aaronontheweb** [comment]
@simonlaroche so that `InsertMany` call will only be used if there are multiple events being persisted at once within the same atomic write - I believe that only happens when a persistent actor calls `PersistAll`. Otherwise all transactions in the array are each done independently.

**Comment by Aaronontheweb** [comment]
Ah, let me rephrase that - the `InsertMany` call will only _send_ multiple events at once if they all come from the same `PersistAll` or `PersistAllAsync` command from one `PersistentActor`. Otherwise, all of the `InsertMany` calls will only contain a single message in it.

**Comment by simonlaroche** [comment]
> Ah, let me rephrase that - the `InsertMany` call will only _send_ multiple events at once if they all come from the same `PersistAll` or `PersistAllAsync` command from one `PersistentActor`. Otherwise, all of the `InsertMany` calls will only contain a single message in it.

Yes.

Let's say you do a persist all with 2 events.

If you do not use a transaction and an error occurs on the second document when doing an insert many, you should get a persistence failure but the first event will be stored.

**Comment by Aaronontheweb** [comment]
Good point - we probably need to fix that.
